usr/opt/OSC200/usr/share/man/man7/pty.7
.\" This manpage source uses rsml coding. 
.so /usr/share/lib/tmac/sml
.so /usr/share/lib/tmac/rsml
.\"
.\"
.\" *********************************************************************
.\" *                                                                   *
.\" *      Copyright (c) Digital Equipment Corporation, 1991, 1994      *
.\" *                                                                   *
.\" *                       All Rights Reserved.                        *
.\" *                                                                   *
.\" *********************************************************************
.\"
.\"
.\" HISTORY
.\"
.\" @(#)$RCSfile: pty.7,v $ $Revision: 3.2.6.4 $ (DEC) $Date: 1993/11/09 19:47:13 $
.\" 
...\"
...\" (c) Copyright 1991, OPEN SOFTWARE FOUNDATION, INC.
...\" ALL RIGHTS RESERVED
...\"
...\"
...\"
...\" Copyright (c) 1983 Regents of the University of California.
...\" All rights reserved.  The Berkeley software License Agreement
...\" specifies the terms and conditions for redistribution.
...\"
...\" @(#)pty.4	6.2 (Berkeley) 5/19/86
...\"
.TH pty 7
.SH NAME
.PP
\*Lpty\*O \- Pseudo terminal driver
.SH SYNOPSIS
.PP
.iX "pty" "pseudo terminal driver"
.iX "pseudo terminal driver"
.iX "drivers" "pseudo terminal"
.sS
\*Lpseudo-device pty [\*Vcount\*O \*L]\*O
.sE
.SH DESCRIPTION
.PP
The \*Lpty\*O driver provides support for a device-pair termed a
\*Lpseudo terminal\*O.  A pseudo terminal is a pair of character devices, a
\*Lmaster\*O device and a \*Lslave\*O device.  The slave device provides
an interface identical to that described in the \*Ltty\*O(7) reference page.
However, whereas all other devices which provide the
interface described in the \*Ltty\*O reference page
have a hardware device behind them, the slave
device has, instead, another process manipulating
it through the master half of the pseudo terminal.
That is, anything written on the master device is
given to the slave device as input and anything written
on the slave device is presented as input on the master device.
.PP
The
DEC OSF/1
operating system
supports a STREAMS-based and \*Lclist\*O-based implementation of the \*Lpty\*O subsystem.
The default configuration uses STREAMS-based \*Lpty\*Os.
STREAMS-based \*Lpty\*Os use the pseudo-device \*Lrpty\*O, while \*Lclist\*O-based \*Lpty\*Os
use the pseudo-device \*Lpty\*O.
By default 32 pseudo terminal special device files are created.
.PP
Note that you cannot have both types of \*Lpty\*Os configured at the
same time.
.PP
To enhance compatibility, STREAMS-based \*Lpty\*Os offers two master pseudo
terminal drivers, the BSD compatible master and the System V compatible
master.  The BSD master is a non-STREAMS device which interfaces to the
STREAMS-based slave \*Lpty\*O.  The System V master is a STREAMS-based device
which also interfaces to the STREAMS-based slave \*Lpty\*O.  The BSD master
is opened through the cloning device, \*L/dev/ptmx_bsd\*O, and through the master
\*Lpty\*O special files, \*L/dev/ptyXX\*O.  The System V master is opened only through
the cloning device \*L/dev/ptmx\*O.  Currently the BSD master cloning device is
used by the \*Llibc\*O routine \*Lopenpty\*O(3).
.\"
.\"In configuring, if an optional \*Vcount\*O is given in
.\"the specification, that number of pseudo terminal pairs are configured;
.\" Begin Digital-specific information CORRECTION
.\" Changed maximum from 816 to 3162 per QAR 9945.
.\" BUT *** Got mail from Ron Menner saying to change it back
.\" to 816 because "the maximum of 3162 pty will *not* be
.\" documented in Silver.
.\" BUT *** got more mail from Ron Menner saying that the supported number 
.\" had changed to approx. 1500.  So, he suggested that I yank the 
.\" information altogether since it keeps changing.
.\" the maximum is 816.
.\" End Digital-specific information CORRECTION
.\" Begin Digital-specific information
.PP
You should allocate ptys by using the \*Lopenpty\*O(3) function, which
hides the pty name space that will change in the next major operating
system release.
.\" End Digital-specific information
.PP
The following \*Lioctl\*O calls apply only to pseudo terminals:
.VL
.LI "ISPTM"
.iX "ISPTM function"
Returns the \*Ldev_t\*O of the
master file descriptor.  ISPTM is valid only on the master half of the pseudo
terminal, and takes no arguments.
.LI "TIOCSTOP"
.iX "TIOCSTOP function"
Stops output to a terminal (for example, like entering \*L<ctrl-S>\*O).  Takes
no parameter.
.LI "TIOCSTART"
.iX "TIOCSTART function"
Restarts output (stopped by TIOCSTOP or by typing \*L<ctrl-S>\*O).
Takes no parameter.
.LI "TIOCPKT"
.iX "TIOCPKT function"
Enable or disable \*Lpacket\*O mode.  Packet mode is enabled by
specifying (by reference) a nonzero parameter and disabled by
specifying (by reference) a zero parameter.  When applied to the
master side of a pseudo terminal, each subsequent \*Lread(\|)\*O
from the terminal will return data written on the slave part of
the pseudo terminal preceded by a zero byte (symbolically
defined as TIOCPKT_DATA), or a single byte reflecting control
status information.  In the latter case, the byte is an inclusive-OR
of zero or more of the bits:
.VL
.LI "TIOCPKT_FLUSHREAD"
Whenever the read queue for the terminal is flushed.
.LI "TIOCPKT_FLUSHWRITE"
Whenever the write queue for the terminal is flushed.
.LI "TIOCPKT_STOP"
Whenever output to the terminal is stopped by \*L<ctrl-S>\*O.
.LI "TIOCPKT_START"
Whenever output to the terminal is restarted.
.LI "TIOCPKT_DOSTOP"
Whenever \*Vt_stopc\*O is \*L<ctrl-S>\*O and \*Vt_startc\*O is \*L<ctrl-Q>\*O.
.LI "TIOCPKT_NOSTOP"
Whenever the start and stop characters are not \*L<ctrl-S>\*O
and \*L<ctrl-Q>\*O.
.LE
.IP
While this mode is in use, the presence of control status information
to be read from the master side may be detected by a \*Lselect(\|)\*O
for exceptional conditions.
.IP
This mode is used by the \*Lrlogin\*O and \*Lrlogind\*O commands
to implement a remote-echoed, locally \*L<ctrl-S>/<ctrl-Q>\*O flow-controlled
remote login with proper back-flushing of output; it can be
used by other similar programs.
.LI "TIOCUCNTL"
.iX "TIOCUCNTL function"
Enable or disable a mode that allows a small number of simple user \*Lioctl\*O
commands to be passed through the pseudo-terminal,
using a protocol similar to that of TIOCPKT.
The TIOCUCNTL and TIOCPKT modes are mutually exclusive.
This mode is enabled from the master side of a pseudo terminal
by specifying (by reference)
a nonzero parameter and disabled by specifying (by reference)
a zero parameter.  Each subsequent \*Lread(\|)\*O
from the master side will return data written on the slave part of
the pseudo terminal preceded by a zero byte,
or a single byte reflecting a user control operation on the slave side.
A user control command consists of a special \*Lioctl\*O
operation with no data; the command is given as UIOCCMD(\*Vn\*O),
where \*Vn\*O is a number in the range 1-255.
The operation value \*Vn\*O will be received as a single byte on the next
\*Lread(\|)\*O from the master side.
The \*Lioctl\*O UIOCCMD(0) is a no-op that may be used to probe for
the existence of this facility.
As with TIOCPKT mode, command operations may be detected with a \*Lselect(\|)\*O
for exceptional conditions.
.LI "TIOCREMOTE"
.iX "TIOCREMOTE function"
A mode for the master half of a pseudo terminal, independent
of TIOCPKT.  This mode causes input to the pseudo terminal
to be flow controlled and not input edited (regardless of the
terminal mode).  Each write to the control terminal produces
a record boundary for the process reading the terminal.  In
normal usage, a write of data is like the data typed as a line
on the terminal; a write of
0 (zero) bytes is like typing an End-of-File
character.  The TIOCREMOTE mode can be used when doing remote line
editing in a window manager, or whenever flow controlled input is required.
.LI "UNLKPT"
.iX "UNLKPT function"
Allows the open
of the corresponding slave to succeed when using the System V master.
If UNLKPT is not used in conjunction with the System V master the open of the
corresponding slave will fail with EPERM.  This ioctl takes
no arguments.
.LE
.SH FILES
.PP
.VL
.\" Begin Digital-specific information
.LI "\*L/dev/pty[a-c, e-z, A-Z][0-9a-f]\*O"
Master pseudo terminals
.LI "\*L/dev/tty[a-c, e-z, A-Z][0-9a-f]\*O"
Slave pseudo terminals
.LI "\*L/dev/ptmx\*O"
System V master cloning device
.LI "\*L/dev/ptmx_bsd\*O"
BSD master cloning device
.\" End Digital-specific information
.iX "pty" "pseudo terminal driver"
.LE
.SH RELATED INFORMATION
.PP
\*Lopenpty\*O(3)
.br
\*ESystem Administration\*O
.EQ
delim off
.EN


usr/opt/OSC200/src/usr/ccs/lib/libc/pty.c
/*
 * *****************************************************************
 * *                                                               *
 * *    Copyright (c) Digital Equipment Corporation, 1991, 1994    *
 * *                                                               *
 * *   All Rights Reserved.  Unpublished rights  reserved  under   *
 * *   the copyright laws of the United States.                    *
 * *                                                               *
 * *   The software contained on this media  is  proprietary  to   *
 * *   and  embodies  the  confidential  technology  of  Digital   *
 * *   Equipment Corporation.  Possession, use,  duplication  or   *
 * *   dissemination of the software and media is authorized only  *
 * *   pursuant to a valid written license from Digital Equipment  *
 * *   Corporation.                                                *
 * *                                                               *
 * *   RESTRICTED RIGHTS LEGEND   Use, duplication, or disclosure  *
 * *   by the U.S. Government is subject to restrictions  as  set  *
 * *   forth in Subparagraph (c)(1)(ii)  of  DFARS  252.227-7013,  *
 * *   or  in  FAR 52.227-19, as applicable.                       *
 * *                                                               *
 * *****************************************************************
 */
/*
 * HISTORY
 */
#ifndef lint
static char *rcsid = "@(#)$RCSfile: pty.c,v $ $Revision: 1.1.12.4 $ (DEC) $Date: 1993/06/07 18:05:51 $";
#endif
/*
 */
/*
 * (c) Copyright 1990, OPEN SOFTWARE FOUNDATION, INC.
 * ALL RIGHTS RESERVED
 */
/*
 * OSF/1 Release 1.0
 */
#if !defined(lint) && !defined(_NOIDENT)

#endif
/*-
 * Copyright (c) 1990 The Regents of the University of California.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms are permitted provided
 * that: (1) source distributions retain this entire copyright notice and
 * comment, and (2) distributions including binaries display the following
 * acknowledgement:  ``This product includes software developed by the
 * University of California, Berkeley and its contributors'' in the
 * documentation or other materials provided with the distribution and in
 * all advertising materials mentioning features or use of this software.
 * Neither the name of the University nor the names of its contributors may
 * be used to endorse or promote products derived from this software without
 * specific prior written permission.
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
 */

#if defined(LIBC_SCCS) && !defined(lint)

#endif /* LIBC_SCCS and not lint */

/* name space pollution clean up */
#ifdef _NAME_SPACE_WEAK_STRONG
#include "pollution.h"
#pragma weak fix_errno = __fix_errno
#pragma weak fork_slvmod = __fork_slvmod
#pragma weak forkpty = __forkpty
#pragma weak openpty = __openpty
#endif
#include <sys/file.h>
#include <sys/ioctl.h>
#include <sys/wait.h>
#include <signal.h>
#include <termios.h>
#include <grp.h>
#include <errno.h>
#ifdef _THREAD_SAFE
#include <sia.h>
#endif

/* constant for max advertised or documented pty supported */
/* (versus number of ptys supported internally - 3162)     */
#define MAX_PTY_ADV 816

int
openpty(amaster, aslave, name, termp, winp)
	int *amaster, *aslave;
	char *name;
	struct termios *termp;
	struct winsize *winp;
{
	register char *clone = "/dev/ptmx_bsd";
	char *template = "/dev/ttyXX";
	char line [10];
	char *cp1 = "pqrstuvwxyzabcefghijklmnoABCDEFGHIJKLMNOPQRSTUVWXYZ";
	char *cp2 = "0123456789abcdef";
	char *cp3 = "ghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
	char *grp_nam = "tty";
#ifdef _THREAD_SAFE
	char buffer[SIABUFSIZ+1];
	int len = SIABUFSIZ;
	struct group _group;
#endif
	char char_8, char_9;
	register int master, slave, euid, ruid, ttygid;
	dev_t  dev;
	struct group *gr;
	void fix_errno();
	
	strcpy(line,template);
#ifdef _THREAD_SAFE
	gr = &_group;
	if (getgrnam_r(grp_nam,gr,buffer,len) == 0)
#else
	if ((gr = getgrnam("tty")) != NULL)
#endif
		ttygid = gr->gr_gid;
	else
		ttygid = -1;
	ruid = getuid();
	euid = geteuid();

	if (( master = open(clone,O_RDWR, 0)) == -1) {
		fix_errno();
		return (-1);	/* out of ptys */
	}
	dev = (dev_t)ioctl(master,ISPTM,0);
	if (dev < 0) {
		(void) close(master);
		fix_errno();
		return (-1);
	}
	if (minor(dev) < MAX_PTY_ADV) {
		char_8 = cp1[(minor(dev))/16];	
		char_9 = cp2[(minor(dev))%16];
	}
	else {
		char_8 = cp1[(minor(dev) - MAX_PTY_ADV)/46];	
		char_9 = cp3[(minor(dev) - MAX_PTY_ADV)%46];
	}
	line[8] = char_8;
	line[9] = char_9;
	if (!euid) {  
		(void) chown(line, ruid, ttygid);
		(void) chmod(line, 0620);
		(void) revoke(line);
	} else {
		if (fork_slvmod(master) < 0) {
			(void) close(master);
			fix_errno();
			return(-1);
		}
	}
	if ((slave = open(line, O_RDWR, 0)) != -1) {
		*amaster = master;
		*aslave = slave;
		if (name)
			strcpy(name, line);
		if (termp)
			(void) tcsetattr(slave, 
				TCSAFLUSH, termp);
		if (winp)
			(void) ioctl(slave, TIOCSWINSZ, 
				(char *)winp);
		return (0);
	}
	(void) close(master);
	fix_errno();
	return (-1);
}

int
fork_slvmod(mast)
register int mast;
{
	int	cstat;
	pid_t	cpid;
	int ret_val;
#ifndef _THREAD_SAFE
	sigset_t newset,oldset;
#endif

	/*
         *  block the SIGCHLD signal
	 */
#ifndef _THREAD_SAFE
	sigemptyset(&newset);
	sigaddset(&newset,SIGCHLD);
	if (sigprocmask(SIG_BLOCK,&newset,&oldset) != 0)
		return(-1);
#endif

	if ((cpid = fork()) < 0)
		return(-1);
	if (cpid) {
		if (waitpid(cpid,&cstat,0) == -1)
			return(-1);
		if (WIFEXITED(cstat))  
			ret_val = 0;
		else
			ret_val = -1;
	/*
	 * restore original signal mask
	 */
#ifndef _THREAD_SAFE
		sigprocmask(SIG_SETMASK,&oldset,(sigset_t *)NULL);
#endif
		return(ret_val);
	}
	else  {
		if (dup2(mast,1) == -1)
			exit(-1);
		execl("/usr/lbin/slvmod","slvmod",0);
		exit(-1); /* shouldn't get here!!!! */
	}
}
void
fix_errno()
{
	int l_errno;
#ifdef _THREAD_SAFE
	l_errno = geterrno(); 
#else
	l_errno = errno;
#endif
	
	switch(l_errno){
		case ENXIO:
		case ENOENT:
		case EMFILE:
		case ENFILE:
		case ENOMEM:
		case EAGAIN:
			break;
		default:
#ifdef _THREAD_SAFE
			seterrno(ENXIO);
#else
			errno = ENXIO;
#endif
	}
}

pid_t
forkpty(amaster, name, termp, winp)
	int *amaster;
	char *name;
	struct termios *termp;
	struct winsize *winp;
{
	int	master, slave;
	pid_t	pid;

	if (openpty(&master, &slave, name, termp, winp) == -1)
		return (-1);
	switch (pid = fork()) {
	case -1:
		fix_errno();
		return (-1);
	case 0:
		/* 
		 * child
		 */
		(void) close(master);
		login_tty(slave);
		return (0);
	}
	/*
	 * parent
	 */
	*amaster = master;
	(void) close(slave);
	return (pid);
}
